fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -0,0 +1,2568 @@
+        final AtomicReference<LeafNode> leafNode = new AtomicReference<>(null);
+
+        return AsyncUtil.whileTrue(() -> {
+            final ChildSlot nextChildSlot = resolveNextIdForFetch(toBeProcessed, mbrPredicate);
+            if (nextChildSlot == null) {
+                return AsyncUtil.READY_FALSE;
+            }
+
+            // fetch the left-most path rooted at the current child to its left-most leaf and concatenate the paths
+            return fetchLeftmostPathToLeaf(readTransaction, nextChildSlot.getChildId(), mbrPredicate).thenApply(nestedTraversalState -> {
+                if (nestedTraversalState.isEnd()) {
+                    // no more data in this subtree
+                    return true;
+                }
+                // combine the traversal states
+                leafNode.set(nestedTraversalState.getCurrentLeafNode());
+                toBeProcessed.addAll(nestedTraversalState.getToBeProcessed());
+                return false;
+            });
+        }, executor).thenApply(v -> leafNode.get() == null
+                                    ? TraversalState.end()
+                                    : TraversalState.of(toBeProcessed, leafNode.get()));
+    }
+
+    /**
+     * Return the next {@link ChildSlot} that needs to be processed given a list of deques that need to be processed
+     * as part of the current scan.
+     * @param toBeProcessed list of deques
+     * @param mbrPredicate a predicate on an mbr {@link Rectangle}
+     * @return The next child slot that needs to be processed or {@code null} if there is no next child slot.
+     *         As a side effect of calling this method the child slot is removed from {@code toBeProcessed}.
+     */
+    @Nullable
+    private static ChildSlot resolveNextIdForFetch(@Nonnull final List<Deque<ChildSlot>> toBeProcessed,
+                                                   @Nonnull final Predicate<Rectangle> mbrPredicate) {
+        for (int level = toBeProcessed.size() - 1; level >= 0; level--) {
+            final Deque<ChildSlot> toBeProcessedThisLevel = toBeProcessed.get(level);
+
+            while (!toBeProcessedThisLevel.isEmpty()) {
+                final ChildSlot nextChild = toBeProcessedThisLevel.pollFirst();
+                if (mbrPredicate.test(nextChild.getMbr())) {
+                    toBeProcessed.subList(level + 1, toBeProcessed.size()).clear();
+                    return nextChild;
+                }
+            }
+        }
+        return null;
+    }
+
+    //
+    // Insert/Update path
+    //
+
+
+    /**
+     * Method to insert an object/item into the R-tree. The item is treated unique per its point in space as well as its
+     * additional key that is also passed in. The Hilbert value of the point is passed in as to allow the caller to
+     * compute Hilbert values themselves. There is, however, that there is a bijective mapping between point and Hilbert
+     * value.
```

Comment:
Something happened to this sentence. I am therefore not clear whether the point is that there is such a bijection, so that the HV argument is somewhat redundant, or that this API design makes it so that there is no actual reliance on such a bijection, so that at least theoretically you could optimize ordering and clustering for something different than location on the Hilbert curve (provided it obeyed some invariants that I am not sure I could list).

Reply:
fixed: 
```
Note that there is a bijective mapping between point and Hilbert
value which allows us to recompute point from Hilbert value as well as Hilbert value from point. We currently
treat point and Hilbert value independent, however, they are redundant and not independent at all. The implication
is that we do not have to store both point and Hilbert value (but we currently do).
 ```